packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -131,34 +146,406 @@ distributionUrl=https\\://services.gradle.org/distributions/gradle-$gradleVersio
+    final RegExp distributionUrlRegex = RegExp(r'distributionUrl=.*\.zip');
+
+    final RegExpMatch? distributionUrl =
+        distributionUrlRegex.firstMatch(wrapperFileContent);
+    if (distributionUrl != null) {
+      // Expected content: 'gradle-7.4.2-all.zip'
+      final String? gradleZip = distributionUrl.group(0);
+      if (gradleZip != null) {
+        final String gradleVersion = gradleZip.split('-')[1];
+        return gradleVersion;
+      } else {
+        // Did not find gradle zip url. Likley this is a bug in our parsing.
+        logger.printWarning(_formatParseWarning(wrapperFileContent));
+      }
+    } else {
+      // If no distributionUrl log then treat as if there was no propertiesFile.
+      logger.printTrace(
+          '$propertiesFile does not provide an Gradle version falling back to system gradle.');
+    }
+  } else {
+    // Could not find properties file.
+    logger.printTrace(
+        '$propertiesFile does not exist falling back to system gradle');
+  }
+  // System installed Gradle version.
+  if (processManager.canRun('gradle')) {
+    final String gradleVersionVerbose =
+        (await processManager.run(<String>['gradle', gradleVersionFlag])).stdout
+            as String;
+    // Expected format:
+/*
+
+------------------------------------------------------------
+Gradle 7.6
+------------------------------------------------------------
+
+Build time:   2022-11-25 13:35:10 UTC
+Revision:     daece9dbc5b79370cc8e4fd6fe4b2cd400e150a8
+
+Kotlin:       1.7.10
+Groovy:       3.0.13
+Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
+JVM:          17.0.6 (Homebrew 17.0.6+0)
+OS:           Mac OS X 13.2.1 aarch64
+    */
+    // Observation shows that the version can have 2 or 3 numbers.
+    // Inner parentheticals `(\.\d+)?` denote the optional third value.
+    // Outter parentheticals `Gradle (...)` denote a grouping used to extract
+    // the version number.
+    final RegExp gradleVersionRegex = RegExp(r'Gradle (\d+\.\d+(\.\d+)?)');
+    final RegExpMatch? version =
+        gradleVersionRegex.firstMatch(gradleVersionVerbose);
+    if (version == null) {
+      // Most likley a bug in our parse implementation/regex.
+      logger.printWarning(_formatParseWarning(gradleVersionVerbose));
+      return null;
+    }
+    return version.group(1);
+  } else {
+    logger.printTrace('Could not run system gradle');
+    return null;
+  }
+}
+
+/// Returns the Android Gradle Plugin (AGP) version that the current project
+/// depends on when found, null otherwise.
+///
+/// The Android plugin version is specified in the [build.gradle] file within
+/// the project's Android directory ([androidDirectory]).
+String? getAgpVersion(Directory androidDirectory, Logger logger) {
+  final File buildFile = androidDirectory.childFile('build.gradle');
+  if (!buildFile.existsSync()) {
+    // TODO should this be a warning?
+    logger.printTrace('Can not find build.gradle in $androidDirectory');
```

Comment:
@christopherfujino Do you think this should be a higher level warning/error

Reply:
Same, printTrace looks correct